libs/injection/bbhnet/injection/injection.py

```diff
@@ -0,0 +1,261 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+import logging
+import os
+
+import bilby
+import h5py
+import numpy as np
+import scipy.signal as sig
+from bilby.gw.conversion import convert_to_lal_binary_black_hole_parameters
+from bilby.gw.source import lal_binary_black_hole
+from gwpy.timeseries import TimeSeries
+
+
+def get_snr(data, noise_psd, fs, fmin=20):
+    """ Calculate the waveform SNR given the background noise PSD"""
+
+    data_fd = np.fft.rfft(data) / fs
+    data_freq = np.fft.rfftfreq(len(data)) * fs
+    dfreq = data_freq[1] - data_freq[0]
+
+    noise_psd_interp = noise_psd.interpolate(dfreq)
+    noise_psd_interp[noise_psd_interp == 0] = 1.0
+
+    snr = 4 * np.abs(data_fd) ** 2 / noise_psd_interp.value * dfreq
+    snr = np.sum(snr[fmin <= data_freq])
+    snr = np.sqrt(snr)
+
+    return snr
+
+
+def generate_gw(
+    sample_params,
+    ifo,
+    waveform_generator=None,
+    get_snr=False,
+    noise_psd=None,
+    whiten_fn=None,
+):
+    """Generate gravitational-wave events
+
+    Arguments:
+    - sample_params: dictionary of GW parameters
+    - sample_duration: time duration of each sample
+    - triggers: trigger time (relative to `sample_duration`) of each sample
+    - ifo: interferometer
+    - waveform_generator: bilby.gw.WaveformGenerator with appropriate params
+    - get_snr: return the SNR of each sample
+    - noise_psd: background noise PSD used to calculate SNR the sample
+    - whiten_fn: whiten each sample using the background noise PSD
+    """
+
+    n_sample = len(sample_params["geocent_time"])
+
+    if waveform_generator is None:
+        waveform_generator = bilby.gw.WaveformGenerator(
+            duration=8,
+            sampling_frequency=16384,
+            frequency_domain_source_model=lal_binary_black_hole,
+            parameter_conversion=convert_to_lal_binary_black_hole_parameters,
+            waveform_arguments={
+                "waveform_approximant": "IMRPhenomPv2",
+                "reference_frequency": 50,
+                "minimum_frequency": 20,
+            },
+        )
+
+    sample_rate = waveform_generator.sampling_frequency
+    waveform_duration = waveform_generator.duration
+    waveform_size = int(sample_rate * waveform_duration)
+
+    signals = np.zeros((n_sample, waveform_size))
+    snr = np.zeros(n_sample)
+
+    ifo = bilby.gw.detector.get_empty_interferometer(ifo)
+    for i in range(n_sample):
+        # Get parameter for one signal
+        p = dict()
+        for k, v in sample_params.items():
+            p[k] = v[i]
+        ra, dec, geocent_time, psi = (
+            p["ra"],
+            p["dec"],
+            p["geocent_time"],
+            p["psi"],
+        )
```

Comment:
If you're just using these once for `ifo.antenna_response` and `ifo.time_delay_from_geocenter`, my inclination would be to skip this and just read from the dict during both those function calls (in keeping with the "three strikes you're out heuristic"), but that's primarily a matter a taste. If it makes those function calls look gross, then forget about it.

Reply:
The only reason I bring it up is that this can be a bit confusing and make it appear like this is doing more than saving characters during the function calls below.